harness/determined/core/_log_shipper.py

```diff
@@ -0,0 +1,188 @@
+        *,
+        session: api.Session,
+        trial_id: int,
+        task_id: str,
+        distributed: Optional[core.DistributedContext] = None
+    ) -> None:
+        self._session = session
+        self._trial_id = trial_id
+        self._task_id = task_id
+        self._distributed = distributed
+
+    def start(self) -> "LogShipper":
+        return self
+
+    def close(
+        self,
+        exc_type: Optional[type],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> "LogShipper":
+        return self
+
+    def __enter__(self) -> "LogShipper":
+        return self.start()
+
+    def __exit__(
+        self,
+        exc_type: Optional[type],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> "LogShipper":
+        return self.close(exc_type, exc_val, exc_tb)
+
+
+class ManagedTrialLogShipper(LogShipper):
+    """
+    Managed trials will ship their logs normally via fluentd.
+    """
+
+    pass
+
+
+class _Interceptor:
+    def __init__(self, original_io: TextIO, handler: Callable[[str], None]) -> None:
+        self._original_io = original_io
+        self._handler = handler
+
+    def write(self, data: str) -> int:
+        self._handler(data)
+        return self._original_io.write(data)
+
+    def flush(self) -> None:
+        self._original_io.flush()
+
+    def __getattr__(self, attr: str) -> Any:
+        return getattr(self._original_io, attr)
+
+
+SHIPPER_FLUSH_INTERVAL = 1
+SHIPPER_FAILURE_BACKOFF_SECONDS = 1
+LOG_BATCH_MAX_SIZE = 1000
+SHIP_QUEUE_MAX_SIZE = 3 * LOG_BATCH_MAX_SIZE
+
+
+class ShutdownMessage:
+    pass
+
+
+QueueElement = Union[str, ShutdownMessage]
+
+
+class LogSender(threading.Thread):
+    def __init__(self, session: api.Session, logs_metadata: Dict) -> None:
+        super().__init__()
+        self._queue = queue.Queue(maxsize=SHIP_QUEUE_MAX_SIZE)  # type: queue.Queue[QueueElement]
+        self._logs = collections.deque()  # type: collections.deque[str]
+        self._session = session
+        self._logs_metadata = logs_metadata
+        self._buf = ""
+
+    def write(self, data: str) -> None:
+        self._queue.put(data)
+
+    def close(self) -> None:
+        self._queue.put(ShutdownMessage())
+
+    def _pop_until_deadline(self, deadline: float) -> Iterator[QueueElement]:
+        while True:
+            timeout = deadline - time.time()
+            if timeout <= 0:
+                break
+
+            try:
+                yield self._queue.get(timeout=timeout)
+            except queue.Empty:
+                break
+
+    def run(self) -> None:
+        while True:
+            deadline = time.time() + SHIPPER_FLUSH_INTERVAL
+            for m in self._pop_until_deadline(deadline):
+                if isinstance(m, ShutdownMessage):
+                    self.ship()
+                    return
+
+                self._logs.append(m)
+                if len(self._logs) >= LOG_BATCH_MAX_SIZE:
```

Comment:
maybe i'm missing something, but why is this logic here and also in `ship()`?

Reply:
the one in `run` works on `.write` call level; but then we need to split these into individual lines. so this logic here is supposed to protect against the `.write`s with a ton of lines.